### PR TITLE
feat(daemon): support dynamically setting the logging level at runtime

### PR DESCRIPTION
**Pull request by sminez**
This lets us do the following:
```bash
$ prompting-client.logging-level --filter=debug
```

and have the daemon update its logging filter without needing to restart the service. By default we will now log at info level.

### CODE REVIEW

File: prompting-client/src/daemon/server.rs [hunk 1]
```diff
@@ -165,6 +209,21 @@ where
...
+        let current = log_filter(&level.into_inner());
+
+        match self.reload_handle.set_filter(&current) {
+            Ok(_) => Ok(Response::new(SetLoggingLevelResponse { current })),
+            Err(e) => Err(Status::new(
+                Code::InvalidArgument,
+                format!("unable to set logging level: {e}"),
+            )),
+        }
+    }
```
Comment (olivercalder) [inline prompting-client/src/daemon/server.rs:226]: Here `level` is passed into `log_filter` as `filter`, but I don't think of levels and filters as really the same thing. What can `level` be set to, and is it validated anywhere?

File: prompting-client/src/lib.rs [hunk 2]
```diff
@@ -11,6 +22,12 @@ mod socket_client;
 mod util;
 
 pub(crate) const SNAP_NAME: &str = "prompting-client";
+pub const SOCKET_ENV_VAR: &str = "PROMPTING_CLIENT_SOCKET";
+pub const DEFAULT_LOG_LEVEL: &str = "info";
+
+pub fn log_filter(filter: &str) -> String {
+    format!("{filter},hyper=error,h2=error")
+}
```
Comment (olivercalder) [inline prompting-client/src/lib.rs:30]: See here.

I would prefer if there were an enum of support log levels, and if the clap arg were defined so the default is explicit.

File: prompting-client/src/protos/apparmor_prompting.rs [hunk 3]
```diff
@@ -417,6 +423,36 @@ pub mod app_armor_prompting_client {
                 );
             self.inner.unary(req, path, codec).await
         }
+        pub async fn set_logging_level(
+            &mut self,
+            request: impl tonic::IntoRequest<::prost::alloc::string::String>,
```
Comment (olivercalder) [inline prompting-client/src/protos/apparmor_prompting.rs:428]: So the logging level is contained in the request?

File: prompting-client/src/daemon/server.rs [hunk 4]
```diff
@@ -42,38 +45,78 @@ macro_rules! map_enum {
...
+        tx_actioned_prompts,
+    );
     let listener = UnixListener::bind(&socket_path).expect("to be able to bind to our socket");
 
     (AppArmorPromptingServer::new(service), listener)
 }
 
-pub struct Service<R>
+pub trait SetLogLevel: Send + Sync + 'static {
+    fn set_filter(&self, level: &str) -> crate::Result<()>;
```
Comment (olivercalder) [inline prompting-client/src/daemon/server.rs:71]: Here as well, we define `set_filter` to take a `level`, but

File: prompting-client/src/daemon/server.rs [hunk 5]
```diff
@@ -42,38 +45,78 @@ macro_rules! map_enum {
...
+}
+
+impl<L, S> SetLogLevel for Arc<Handle<L, S>>
+where
+    L: From<EnvFilter> + Send + Sync + 'static,
+    S: 'static,
+{
+    fn set_filter(&self, filter: &str) -> crate::Result<()> {
```
Comment (olivercalder) [inline prompting-client/src/daemon/server.rs:79]: then `set_filter` is implemented to take a `filter`.

File: prompting-client/src/daemon/server.rs [hunk 6]
```diff
@@ -42,38 +45,78 @@ macro_rules! map_enum {
...
+        info!(?filter, "attempting to update logging filter");
+        let f = filter
+            .parse::<EnvFilter>()
+            .map_err(|_| Error::UnableToUpdateLogLevel {
+                reason: format!("{filter:?} is not a valid logging filter"),
+            })?;
```
Comment (olivercalder) [inline prompting-client/src/daemon/server.rs:85]: It looks like `EnvFilter` is defined in another crate (not in this repo) so I definitely think there should be some documentation (both in the code and in the CLI args) about what valid log levels are.

File: prompting-client/src/daemon/server.rs [hunk 7]
```diff
@@ -42,38 +45,78 @@ macro_rules! map_enum {
...
+
+        self.reload(f).map_err(|e| Error::UnableToUpdateLogLevel {
```
Comment (olivercalder) [inline prompting-client/src/daemon/server.rs:87]: Reload also isn't defined in this repo either? Seems like some `tonic` magic or something

**Review by matthew-hagemann — APPROVED** [review]
:tada:

File: prompting-client/src/bin/set_log_level.rs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+use std::process::exit;
+
+/// Set the logging level for a running instance of the prompting client daemon
+#[derive(Debug, Parser)]
+#[clap(about, long_about = None)]
+struct Args {
+    /// The filter to use for setting the logging level
+    #[clap(short, long, value_name = "FILTER")]
+    level: String,
+}
```
[the remark this answers is not in the document]
Comment (sminez) [reply prompting-client/src/bin/set_log_level.rs:12]: There's not a single list, no. The API docs [here](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html) cover what is possible which really is a _lot_, which is why I've not restricted this to a fixed list of levels like trace, debug, info etc as an enum. That would prevent us using more granular targeting of logs in situations where we really want it.

File: prompting-client/src/daemon/server.rs (hunk 7 above)
Comment (sminez) [reply prompting-client/src/daemon/server.rs:87]: it's a feature of tracing_subscriber, yeah. All of this is just a way of plumbing a valid logging filter through to this call to `reload`

**Comment by sminez** [comment]
@olivercalder I see what you mean about being inconsistent around levels vs filters. I'll update things to consistently use the term `filter` everywhere but documenting the full functionality of what is supported here (or restricting things to a subset) is probably not what we want to go for. I'll update the clap docs so that they cover the common use cases of _just_ setting a logging level for the daemon code itself. The issue is that a very large number of possible filters are possible and I don't really want to link to Rust API documentation in the CLI.

File: prompting-client/src/bin/set_log_level.rs
Comment (olivercalder) [inline prompting-client/src/bin/set_log_level.rs:1]: I think the filename could be updated to match as well

File: prompting-client/src/cli_actions/log_level.rs
Comment (olivercalder) [inline prompting-client/src/cli_actions/log_level.rs:1]: same

File: prompting-client/src/bin/set_log_level.rs [hunk 9]
```diff
@@ -1,22 +1,43 @@
+//! This is a helper command for setting our logging filter within the daemon while it is running.
+//! Doing this involves having to send a GRPC message to the daemon which inolves more plumbing
+//! than we'd ideally like but it allows us to update the filter without having to restart the
+//! daemon.
+//!
+//! The docs on the command itself cover how to set simple logging levels but the tracing framework
+//! we are using actually supports a rich syntax for how these filters are written. As and when we
+//! need to use this in anger, the docs can be found here:
+//!   https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html
```
Comment (olivercalder) [inline prompting-client/src/bin/set_log_level.rs:9]: Thanks, this is great!

File: prompting-client/src/bin/set_log_level.rs [hunk 10]
```diff
@@ -1,22 +1,43 @@
...
+/// to the system journal by the prompting-client daemon. The supported levels are (in order of
+/// least to most verbose):
+///
+///   - error
+///   - warn
+///   - info
+///   - debug
+///   - trace
+///
+/// The default level is "info".
```
Comment (olivercalder) [inline prompting-client/src/bin/set_log_level.rs:26]: Perfect!

**Review by olivercalder — APPROVED** [review]
Looks great, thanks for the changes! Some filenames still use "log level" terminology, and could be updated to use "filter" instead, but not a huge deal.

**Pull request merged**